### PR TITLE
Persist metadata attribute protection level and protect stored content

### PR DESCRIPTION
**Pull request by 3keyroman**
The protection level declared in metadata attribute properties is now persisted on the attribute definition when the definition is created or updated, so metadata content declared as encrypted is stored encrypted at rest. A declared level of null is treated as none.

When the declared protection level diverges from the persisted one, already stored content items of the definition are re-protected accordingly: encrypted on upgrade to encrypted, decrypted on downgrade.

The metadata read projections now carry the content item ciphertext, and encrypted content is decrypted when returning metadata content, mapped metadata, and definition object attribute content.

### CODE REVIEW

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java [hunk 1]
```diff
@@ -670,6 +686,10 @@ public AttributeDefinition updateMetadataAttributeDefinition(MetadataAttribute m
             attributeDefinition = attributeDefinitionRepository.findByTypeAndConnectorUuidAndAttributeUuidAndName(AttributeType.META, connectorUuid, UUID.fromString(metadataAttribute.getUuid()), metadataAttribute.getName()).orElse(null);
         }
         String label = metadataAttribute.getProperties().getLabel();
+        ProtectionLevel protectionLevel = metadataAttribute.getProperties().getProtectionLevel();
+        if (protectionLevel == null) {
+            protectionLevel = ProtectionLevel.NONE;
+        }
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java:743]: Fixed in e50681280: the persisted definition copy now gets a detached properties copy (Jackson convertValue, so future fields survive) carrying the normalized protection level, keeping the serialized definition consistent with the entity column. One note on the failure mode: the mismatch did not force an update on every resubmission — the first divergent write persisted the new serialization, so it converged after one UPDATE. The remaining issue was the stored definition JSON carrying null while the column says NONE, which mapToGlobalMetadataDefinitionDetailDto would surface as a null protection level.

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java [hunk 2]
```diff
@@ -699,6 +723,11 @@ public AttributeDefinition updateMetadataAttributeDefinition(MetadataAttribute m
             attributeDefinition.setVersion(metadataAttribute.getVersion());
             attributeDefinition.setGlobal(isGlobal);
         }
+        // Re-protect stored content when the declared level diverges from the persisted one. This also
+        // heals definitions whose content was stored in plaintext before the metadata protection level
+        // was persisted at all: the unset->ENCRYPTED transition encrypts every existing content item.
+        encryptOrDecryptExistingContent(attributeDefinition, protectionLevel);
+        attributeDefinition.setProtectionLevel(protectionLevel);
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java:782]: Fixed in e50681280: encryptOrDecryptExistingContent now writes the placeholder with attributeDefinition.getVersion() instead of the hard-coded V3. This is behavior-neutral for custom attributes (their definitions are always persisted as V3) and produces a V2-shaped placeholder for V2 metadata definitions, covered by a new test asserting the placeholder content class for a V2 definition upgrade.

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java [hunk 3]
```diff
@@ -699,13 +708,49 @@ public AttributeDefinition updateMetadataAttributeDefinition(MetadataAttribute m
...
+    /**
+     * Returns the content item as stored, or the decrypted item when the definition's protection
+     * level put ciphertext alongside the stored placeholder.
+     */
+    private static AttributeContent decryptedContentItem(AttributeContent contentItem, int version, AttributeContentType contentType, String encryptedContent) {
+        if (encryptedContent == null) {
+            return contentItem;
+        }
+        return AttributeVersionHelper.decryptContent(contentItem, version, contentType, encryptedContent);
+    }
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java]: Fixed in a41b19311: decryptedContentItem catches RuntimeException, logs a warning with the item reference, and falls back to the stored placeholder — the metadata loops then skip it via the existing getData() == null check. Covered by a test that corrupts a stored ciphertext and asserts both metadata reads succeed with the item skipped. The warn log is deliberate so a wrong encryption key does not just manifest as silently empty metadata.

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java [hunk 4]
```diff
@@ -760,14 +810,15 @@ public List<DataAttribute> getDefinitionObjectAttributeContent(AttributeType att
         Map<String, DataAttribute> mapping = new HashMap<>();
         for (ObjectAttributeDefinitionContent objectDefinitionContent : objectDefinitionContents) {
             String uuid = objectDefinitionContent.uuid().toString();
+            AttributeContent contentItem = decryptedContentItem(objectDefinitionContent.contentItem(), objectDefinitionContent.definition().getVersion(), ((DataAttribute) objectDefinitionContent.definition()).getContentType(), objectDefinitionContent.encryptedContent());
             if (objectDefinitionContent.definition().getVersion() == 2) {
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java:887]: Fixed in b8d5d6416: getDefinitionObjectAttributeContent now skips items whose data is null after decryption, matching the malformed-data convention of the metadata read paths. CredentialServiceITest, DiscoveryServiceITest and LocationServiceITest (the consumers of this method) all pass, confirming no legitimate items are dropped.

File: src/test/java/com/otilm/core/integration/attribute/AttributeEngineITest.java [hunk 5]
```diff
@@ -1091,6 +1091,179 @@ private <T> T inNewTransaction(Supplier<T> work) {
...
+                ObjectAttributeContentInfo.builder(Resource.CERTIFICATE, certificate.getUuid()).connector(connectorAuthority.getUuid()).build(),
+                List.of(requestAttribute));
+
+        DataAttribute readAttribute = attributeEngine.getDefinitionObjectAttributeContent(AttributeType.DATA, connectorAuthority.getUuid(), null, Resource.CERTIFICATE, certificate.getUuid()).stream()
+                .filter(a -> a.getName().equals(secretAttribute.getName())).findFirst().orElseThrow();
+        List<AttributeContent> readContent = readAttribute.getContent();
+        Assertions.assertEquals("encrypted-at-rest", readContent.getFirst().getData(), "definition object content read must return the decrypted value");
+    }
+
     @Test
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/test/java/com/otilm/core/integration/attribute/AttributeEngineITest.java:1339]: Added in b8d5d6416: getDefinitionObjectAttributeContentDecryptsEncryptedContent now also corrupts the stored ciphertext and asserts the read succeeds with the undecryptable item skipped rather than returned with null data.

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java [hunk 6]
```diff
@@ -699,13 +708,54 @@ public AttributeDefinition updateMetadataAttributeDefinition(MetadataAttribute m
...
+     * (corrupt value, wrong encryption key) degrades to the stored placeholder — whose {@code data}
+     * is null, so callers treat it as malformed — instead of failing the whole read.
+     */
+    private static AttributeContent decryptedContentItem(AttributeContent contentItem, int version, AttributeContentType contentType, String encryptedContent) {
+        if (encryptedContent == null) {
+            return contentItem;
+        }
+        try {
+            return AttributeVersionHelper.decryptContent(contentItem, version, contentType, encryptedContent);
+        } catch (RuntimeException e) {
```
Comment (lubomirw) [inline src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java]: **[Medium]** | Error handling

`decryptedContentItem` catches every `RuntimeException`, logs a per-item WARN, and returns the null-data placeholder — which all read paths then skip. Two issues: (a) the broad catch masks genuine defects (NPE / ClassCastException) as "decrypt failed"; (b) a rotated/lost content-encryption key or corrupt ciphertext makes protected content silently vanish from reads (a credential/discovery attribute whose only content is undecryptable disappears entirely), with only a per-item WARN — the connector then receives incomplete data and fails downstream with no clear cause. Confidentiality is safe (placeholder data is null, never plaintext); the concern is silent data-loss + hidden bugs.

_Suggested:_ narrow the catch to the specific crypto/parse exception types so unexpected runtime errors still fail loudly; log the full exception plus the definition/object identity (not only `e.getMessage()`); for credential-bearing reads, consider surfacing a distinct unreadable-content error instead of silent omission.

<sub>flagged by Copilot CLI, guided-pr-review, superpowers code-review</sub>

**Review by lubomirw — COMMENTED** [review]
Automated multi-reviewer triage — 3 items (COMMENT). Hygiene clean; 1 inline finding + 2 items below (off-diff / general).

---

**#3 — No metadata downgrade test (ENCRYPTED→NONE)** · Low · Testing · `AttributeEngineITest.java`

The Javadoc and PR description state stored content is decrypted on downgrade, and the upgrade path (NONE→ENCRYPTED) is well covered — but no metadata test exercises the ENCRYPTED→NONE transition decrypting already-stored content and clearing `encryptedData`.

_Suggested:_ add an integration test that registers metadata as ENCRYPTED with stored content, re-registers it as NONE, and asserts the content item becomes plaintext, `encryptedData` is null, and it still reads back correctly.

<sub>flagged by guided-pr-review</sub>

---

**#2 — Question: is encrypted metadata content meant to bypass dedup?** · `AttributeEngine.java:1654-1661` (unchanged by this PR, so noted here rather than inline)

Now that the metadata protection level is persisted, ENCRYPTED metadata content routes through the encrypted write branch, which sets `contentItemEntity = null` and skips both `findByJsonAndAttributeDefinitionUuid` and `findExistingContentMapping`, always inserting a new `AttributeContentItem` + mapping. For encrypted DATA attributes this is pre-existing (ciphertext can't be compared — possibly deliberate), but it is newly reachable for metadata via this change.

Question to confirm: is re-registering the same encrypted metadata on the same object+version with `overwrite=false` expected to accumulate duplicate content rows, or should the encrypted path also dedup (e.g. match on reference, or run the existing duplicate-mapping check before insert)? Not asserting a bug — flagging to confirm the intended behavior.

<sub>flagged by Copilot CLI (low confidence)</sub>

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java (hunk 6 above)
Comment (3keyroman) [reply src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java]: Addressed in 630d22fb0. (a) The catch is narrowed to IllegalArgumentException | IllegalStateException | DateTimeException — the actual failure surface of SecretsUtil decode/decrypt and the typed-value parsing in decryptContent — so NPE/ClassCastException and other unexpected errors propagate again. (b) The warn now logs the full exception plus the attribute name and object identity at every call site, so a rotated/lost key shows up as a clear repeated signal rather than a bare message. On surfacing a distinct unreadable-content error for credential-bearing reads: that pulls against the earlier review round that asked these reads to degrade rather than hard-fail on a single corrupt row; with the narrowed catch and identity-rich WARN the failure is diagnosable, and the definition-level content (credential secrets on the AttributeDefinition itself) is unaffected by this path. If you prefer loud failure specifically for CREDENTIAL resource reads, that is a small follow-up change — say the word.

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java [hunk 7]
```diff
@@ -1641,6 +1706,28 @@ private void createObjectAttributeContent(AttributeDefinition attributeDefinitio
         }
     }
 
+    /**
+     * Locates an already-mapped content item of the definition whose decrypted value equals the
+     * incoming plaintext item. Scoped to the target object tuple so the number of decryptions per
+     * write stays bounded by the object's own content, not by every object sharing the definition.
+     */
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java:1790]: Fixed in 16f852f71: the Javadoc now states the equality precisely — matching data, reference and content type — which is the intended semantics: deduplication only merges exact-value resubmissions and can never merge items that differ in any part of the content.

File: src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java [hunk 8]
```diff
@@ -1596,10 +1656,15 @@ private void createObjectAttributeContent(AttributeDefinition attributeDefinitio
...
-                encryptedData = encryptAttributeContent(attributeDefinition, attributeContentItem);
-                attributeContentItem = AttributeVersionHelper.createEncryptedContent(attributeContentItem.getReference(), attributeDefinition.getContentType(), attributeDefinition.getVersion());
+                // Encrypted content cannot be deduplicated by its stored json (the placeholder is
+                // identical for every item and the ciphertext is salted) — dedup by decrypting the
+                // items already mapped to this object tuple and comparing plaintext.
+                contentItemEntity = findMappedEncryptedContentItem(attributeDefinition, attributeContentItem, objectAttributeContentInfo);
+                if (contentItemEntity == null) {
+                    encryptedData = encryptAttributeContent(attributeDefinition, attributeContentItem);
+                    attributeContentItem = AttributeVersionHelper.createEncryptedContent(attributeContentItem.getReference(), attributeDefinition.getContentType(), attributeDefinition.getVersion());
+                }
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/attribute/engine/AttributeEngine.java:1738]: Fixed in c601b84bd: the mapped items are now fetched and decrypted once per write via loadDecryptedMappedItems — one query and M decryptions per batch instead of N queries and N*M decryptions — and the per-item dedup is an in-memory comparison. Items inserted during the batch are appended to the decrypted list so equal values later in the same content list still deduplicate.

**Review by lubomirw — APPROVED** [review]
